aidial_adapter_openai/utils/globals.py

```diff
@@ -0,0 +1,17 @@
+import functools
+
+import httpx
+
+from aidial_adapter_openai.constant import (
+    DEFAULT_CONNECTION_LIMITS,
+    DEFAULT_TIMEOUT,
+)
+
+
+@functools.lru_cache(maxsize=1)
```

Comment:
What do you think, if we make it this way?

```python
class GlobalAsyncClient:
    _client: httpx.AsyncClient | None

    def __init__(self):
        self._client = None

    def set_up_client(self):
        self._client = httpx.AsyncClient(
            timeout=DEFAULT_TIMEOUT,
            limits=DEFAULT_CONNECTION_LIMITS,
            follow_redirects=True,
        )

    async def tear_down_client(self):
        if self._client is None:
            raise Exception("Trying to tear down client before set up!")
        await self._client.aclose()

    @property
    def client(self) -> httpx.AsyncClient:
        if self._client is None:
            raise Exception("Trying to get no setted client!")
        return self._client
```

Reply:
So then we could make explicit set up and tear down in lifespan:

```python
global_async_client = GlobalAsyncClient()

@asynccontextmanager
async def lifespan(app: FastAPI):
    global_async_client.set_up_client()
    yield
    logger.info("Application shutdown")
    await global_async_client.tear_down_client()
```

Reply:
And that's not everything - I need to implement a singleton class, because I need to use the instance in other modules too: `.mistral` and `.utils.parsers`.

Too much code for little value. The only thing we gain here is that `lifespan` is more readable and maintainable.
But we loose clarity and maintainability in all other places.

Let's keep it simple.

Reply:
Wait, but if you just make this in `globals.py`

```
global_async_client = GlobalAsyncClient()
```
how is it messing with `mistral` and `utils`?


---

The main concern I have, is that when instead of object we use this `@functools.cache`:

1. It's really misleading, what we're trying to achieve, because we don't want to cache something, we want one singleton client for all requests
2. I haven't used `functools.cache` for this singleton approach not once in my life. I can't guarantee it don't have some side effects